### ISSUE TITLE
Turtles causing server lag and eventual crash

### ISSUE THREAD

**Issue by hugeblank** [SquidDev-CC/CC-Tweaks#139]
I have no idea how this issue is occuring, but project beta really likes to crash.
Judging from the timings report that I managed to pull before the server went down, turtles are to blame.
Here's the timings report: https://timings.aikar.co/v2/?id=9ef938801a454770a5e9cc0e907b805e&section=lagsummary

Server info:
Sponge: 2006
Forge: 2185

Is this your mod, or is it something else? It's been happening rather frequently and is a nuisance to the players.

**Comment by inklit** [comment]
If it happens again, post a WarmRoast.

**Comment by SquidDev** [comment]
As far as I'm aware this isn't CCTweaks' fault: I don't fiddle that much with turtles. However, it is something I'm interested in fixing. 

I was wondering if, next time this happens, you could run a profiler like [jvmtop](https://github.com/patric-r/jvmtop/blob/master/doc/ConsoleProfiler.md) or WarmRoast and also provide the output of [`jstack`](http://docs.oracle.com/javase/7/docs/technotes/tools/share/jstack.html). (IIRC the former requires the JDK so if you don't have it installed then don't worry).

From from talking with @inklit, it sounds like this is someone deliberately causing lag.

**Comment by hugeblank** [comment]
you think so? I don't know anyone on our usernamecache that would have the balls to do that 😆
and also, _is that even possible???_ I can't think of any way to crash a server without using a boatload of computers, or in this case, turtles. The server really hasn't been up long enough for anyone to reach that kind of ability. Is there a way to kill it by using any less than like 30?

As far as profilers go, I would have to talk to Danny about it. Knowing him, I'm sure he has some sort of one on the server already.

EDIT: maybe you would want the crash report too?

Expect results by Monday :P

**Comment by inklit** [comment]
@hugeblank 

Let me explain to you how the Sponge aikar timing system works, it is not your traditional profiler.
This is only loosely speaking and likely isn't entirely accurate, but I hope you get the idea.

## The Aikar Timing System
First, while the server does specific operations, it will make a call to the Aikar timing system to start recording the current time (in nanoseconds or ticks), and then it will do an operation, and after the operation is completed it will make another call to the Aikar timing system to stop recording the time.
The delta in time is then recorded in the profiler which is what you see in your timings report.

_This means that, in order for a Turtle tile entity to actually increase in time, it must be interacting with your tick thread by making calls that change the world in some way._

This may be a quite wild jump, but let me explain the logic behind that...Lua code runs on a completely different thread than the Minecraft server threads themselves. 
This means Aikar can't profile things like Lua operations even though you can cause severe server lag by abusing Lua functions.
 Any lag caused by a Computer or Turtle that is doing something like spamming large tables to exhaust memory will not show in the Aikar timing system, but what will show on the Aikar timing system is a vague "TPS loss" that is not directly linked to anything that has been profiled. 
The profiler only realizes that we're losing TPS, however it can't explain why.

## So what's going on?
In your timing report, it shows a clear explanation as to why the lag is caused and that is the ticking of the Turtle tile entity. This means there is something (maybe called by the Lua engine or not) that is causing the Turtle tile entity tick to either take an excessive amount of time to tick, or delaying the ticks in some way that Aikar is capable of profiling, however the profiler does not get much more detailed other than the fact that it's just "ticking".

This has been seen in various situations on my server and almost all of them were caused by Turtles interacting with the world. Here is a list of things I've personally seen: Rapid Redstone interactions, rapid chest interactions, rapid item picking up and dropping, a sheer large amount of Turtles moving in general, rapidly placing and destroying on large scales. Most of these were not players intentionally trying to cause lag, but having rather large tick involved projects.

## More analysis
I'd like to point out this part of your timings report: http://i.imgur.com/tqZHpo8.png
This is pretty suspicious that it's accounting for 32%, so maybe you should look into Turtles that are dealing with Redstone.

Of course, this does not eliminate the possibility of there being a silly bug in the code base that is causing the Turtle tile entity to tick excessively long, but we cannot be sure without a real profiler showing what function calls are taking up the most time.

## More details needed
The timing report really does not tell anybody much except that Turtle's tile entities are taking a wild amount of time to tick. It doesn't show what functions are being called or how long they are taking.

By the timing report alone we cannot tell whether or not it's a function that is being called that is just slow, or if it is thousands of functions being called that are fast but add up to be slow.  


The profiler results are crucial for figuring out why this is happening.


### tl;dr get profiler

**Comment by hugeblank** [comment]
I'm pretty sure I know what the repeaters in dimension -1 were. In the nether I built a farm for blazes that uses a significant amount of redstone. So at the time, it would make sense that it was causing lag, it's got some clocking redstone that at that time would have been loaded. 

Of course me saying that doesn't completely factor out the possibility of there being a turtle in the nether killing my server by ticking a redstone repeater, but worth mentioning.

will get profiler :P

**Comment by SquidDev** [comment]
I'm going to close this for now. This doesn't appear to be a CCTweaks but and so isn't an immediate priority. If you can provide some profiling information I might be able to have a go at resolving some of these issues.

**Comment by hugeblank** [comment]
alright, I'm fine with that.
